### PR TITLE
Add stream buffering for named pipe reads

### PR DESCRIPTION
**Pull request by dsplaisted**
This change adds a read buffer for reads from the named pipe streams which are used to communicate between MSBuild processes.  As there are lots of short reads in the serialization code, this should improve performance of the communication.

### CODE REVIEW

**Comment by AndyGerlicher** [comment]
@dsplaisted did you change something that fixed the unit test? Looked like a legit failure (null ref so the underlying stream was null). Approach looks good though in general. Only thing I wish is we didn't have to have our own copy of that class. Did you copy this from somewhere? I saw there was one in coreclr, but it didn't seem to be public :(.

Were you able to verify this was any faster?

Thanks a ton for looking into this!

**Comment by dsplaisted** [comment]
Yes, I fixed the unit test failure.

I didn't copy the code from anywhere, because the various buffered stream implementations I saw in .NET didn't exactly meet the requirements here.  They assumed that reads and writes to the stream applied to the same set of data, so writing to the stream would invalidate the read buffer, and vice versa.  The pipe streams we use are for duplex communication, so there's no relation between the reads and writes to the stream.  So the buffered stream I wrote buffers reads, but sends writes immediately to the underlying stream.

This sped up the scenario I was testing from around 70+ seconds to between 25 and 40 seconds.  There may be more perf to be gained.  This change only buffers the reads in one direction, and it would seem that also buffering in the other direction could help more but in my testing it slowed it down significantly (or maybe just hung, I stopped waiting after several minutes).

File: src/Shared/BufferedReadStream.cs [hunk 1]
```diff
@@ -0,0 +1,152 @@
...
+                {
+                    Array.Copy(_buffer, _currentIndexInBuffer, buffer, offset, _remainingBufferSize);
+                    alreadyCopied = _remainingBufferSize;
+                    _currentIndexInBuffer = 0;
+                    _remainingBufferSize = 0;
+                }
+
+                int innerReadCount = _innerStream.Read(_buffer, 0, BUFFER_SIZE);
+                _currentIndexInBuffer = 0;
+                _remainingBufferSize = innerReadCount;
```
Comment (rainersigwald) [inline src/Shared/BufferedReadStream.cs]: Shouldn't this be `BUFFER_SIZE - innerReadCount`?

File: src/Shared/BufferedReadStream.cs [hunk 2]
```diff
@@ -0,0 +1,152 @@
...
+            _remainingBufferSize = 0;
+        }
+
+        public override bool CanRead { get { return _innerStream.CanRead; } }
+
+        public override bool CanSeek { get { return false; } }
+
+        public override bool CanWrite { get { return _innerStream.CanWrite; } }
+
+        public override long Length { get { return _innerStream.Length; } }
```
Comment (cdmihai) [inline src/Shared/BufferedReadStream.cs]: Isn't this _innerStream.Length + _currentIndexInBuffer?

File: src/Shared/BufferedReadStream.cs (hunk 1 above)
Comment (dsplaisted) [inline src/Shared/BufferedReadStream.cs]: This is the number of bytes from the underlying stream that have been read into the buffer but not by consumers of the buffered stream.  I've renamed the field to `_currentlyBufferedByteCount` to make this more clear.

File: src/Shared/BufferedReadStream.cs [hunk 3]
```diff
@@ -0,0 +1,152 @@
...
+
+        public override long Position
+        {
+            get { return _innerStream.Position; }
+            set { _innerStream.Position = value; }
```
Comment (dsplaisted) [inline src/Shared/BufferedReadStream.cs]: True.  This property shouldn't really be used for this scenario anyway, so I've changed it to throw a `NotSupportedException`.

**Comment by dsplaisted** [comment]
> Overall looks good but I'd really like to see some tests for this.

This does get tested with tests that do multiprocess builds, so I'm assuming you mean unit-level tests.  Yeah... that would be nice... :-)

**Comment by jeffkl** [comment]
In my perf testing, I didn't see any issues with the change but it didn't seem to be any faster either.

**Comment by rainersigwald** [comment]
Ok, let's get this in for some mileage in RC. Tests should come ASAP #1272.

**Pull request merged**